File: mapstring.php
<?php
/**
 * Maps a string looking for JSON patterns
 * @param string $str Input string
 * @return array key => values and value => keys
 */

function mapstring($str) {
	$all = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.';
	$len = strlen($str);
	$key = '';
	$val = '';
	$qut = '';
	$run = 0;
	
	for ($i = 0; $i < $len; $i++) {
		$chr = $str{$i};
		
		// quoted text
		if ($qut) {
			// quotes ended
			if ($chr == $qut) {
				$qut = '';
			}
			else {
				$val .= $chr;
			}
		}
		// open quotes
		else if ($chr == '"' || $chr == "'") {
			$qut = $chr;
		}
		// make
		else if (strpos(',]};', $chr) !== false) {
			// format the value
			if ($key) {
				echo "$key => $val\n";
			}
			$key = '';
			$val = '';
			$run = 0;
		}
		// key value switcher
		else if (strpos(':=', $chr) !== false) {
			$key = $val;
			$val = '';
		}
		// new divisor
		else if (strpos('{[', $chr) !== false) {
			$key = '';
			$val = '';
		}
		// letters, numbers and dots
		else if (stripos($all, $chr) !== false) {
			$val .= $chr;
		}
	}
}

File: mapstring.cpp
#include <iostream>

void mapstring(std::string str) {
	std::string all =
		"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
		"abcdefghijklmnopqrstuvwxyz"
		"0123456789.";
	std::string key = "";
	std::string val = "";

	bool mke = false;
	char qut = '\0';

	for (char& chr: str) {
		// quoted text
		if (qut != '\0') {
			if (chr == qut) {
				qut = '\0';
			}
			else {
				val += chr;
			}
		}
		// quotes found
		else if (chr == '\'' || chr == '"') {
			qut = chr;
		}
		// make
		else if (chr == ',' || chr == ']' || chr == '}' || chr == ';') {
			if (!key.empty()) {
				std::cout << key << " => " << val << std::endl;
			}
			key = "";
			val = "";
		}
		// key value switcher
		else if (chr == ':' || chr == '=') {
			key = val;
			val = "";
		}
		// new block
		else if (chr == '{' || chr == '[') {
			key = "";
			val = "";
		}
		// letters, numbers and dots
		else if (std::find(std::begin(all), std::end(all), chr) != std::end(all)) {
			val += chr;
		}
	}
}
